sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java

```diff
@@ -92,67 +88,23 @@
-      MessageType fileSchema = footer.getFileMetaData().getSchema();
-      FilterCompat.Filter filter = getFilter(configuration);
-      blocks = filterRowGroups(filter, footer.getBlocks(), fileSchema);
-    } else {
-      // SPARK-33532: After SPARK-13883 and SPARK-13989, the parquet read process will
-      // no longer enter this branch because `ParquetInputSplit` only be constructed in
-      // `ParquetFileFormat.buildReaderWithPartitionValues` and
-      // `ParquetPartitionReaderFactory.buildReaderBase` method,
-      // and the `rowGroupOffsets` in `ParquetInputSplit` set to null explicitly.
-      // We didn't delete this branch because PARQUET-131 wanted to move this to the
-      // parquet-mr project.
-      // otherwise we find the row groups that were selected on the client
-      footer = readFooter(configuration, file, NO_FILTER);
-      Set<Long> offsets = new HashSet<>();
-      for (long offset : rowGroupOffsets) {
-        offsets.add(offset);
-      }
-      blocks = new ArrayList<>();
-      for (BlockMetaData block : footer.getBlocks()) {
-        if (offsets.contains(block.getStartingPos())) {
-          blocks.add(block);
-        }
-      }
-      // verify we found them all
-      if (blocks.size() != rowGroupOffsets.length) {
-        long[] foundRowGroupOffsets = new long[footer.getBlocks().size()];
-        for (int i = 0; i < foundRowGroupOffsets.length; i++) {
-          foundRowGroupOffsets[i] = footer.getBlocks().get(i).getStartingPos();
-        }
-        // this should never happen.
-        // provide a good error message in case there's a bug
-        throw new IllegalStateException(
-            "All the offsets listed in the split should be found in the file."
-                + " expected: " + Arrays.toString(rowGroupOffsets)
-                + " found: " + blocks
-                + " out of: " + Arrays.toString(foundRowGroupOffsets)
-                + " in range " + split.getStart() + ", " + split.getEnd());
-      }
-    }
-    this.fileSchema = footer.getFileMetaData().getSchema();
-    Map<String, String> fileMetadata = footer.getFileMetaData().getKeyValueMetaData();
+    ParquetReadOptions options = HadoopReadOptions
+      .builder(configuration)
+      .withRange(split.getStart(), split.getStart() + split.getLength())
+      .build();
+    this.reader = new ParquetFileReader(HadoopInputFile.fromPath(file, configuration), options);
+    this.fileSchema = reader.getFileMetaData().getSchema();
+    Map<String, String> fileMetadata = reader.getFileMetaData().getKeyValueMetaData();
     ReadSupport<T> readSupport = getReadSupportInstance(getReadSupportClass(configuration));
     ReadSupport.ReadContext readContext = readSupport.init(new InitContext(
         taskAttemptContext.getConfiguration(), toSetMultiMap(fileMetadata), fileSchema));
     this.requestedSchema = readContext.getRequestedSchema();
     String sparkRequestedSchemaString =
```

Comment:
Maybe we should call `this.reader.setRequestedSchema(this.requestedSchema);` at there.

Reply:
It seems that the new API initializes `ParquetFileReader ` using `fileSchema` not `requestedSchema `

Reply:
New API:
![image](https://user-images.githubusercontent.com/1475305/108979649-d5a61b80-76c5-11eb-88d6-b0acce5c4bec.png)

Old API:
![image](https://user-images.githubusercontent.com/1475305/108979918-1b62e400-76c6-11eb-8491-6fccd9bdd33a.png)

Reply:
Nice find @LuciferYang ! I think you're right. The `requestedSchema` was used to initialize `reader` but not in the new code path. This could also explain why we didn't see perf regression in `FilterPushdownBenchmark` since it doesn't have  any column projection. I'll try the fix with TPCDSQueryBenchmark and report result here later.